clos.md

```diff
@@ -1542,20 +1542,64 @@ It's working.
 
 ## Controlling the initialization of instances (initialize-instance)
 
-To further customize the creation of instances by specializing
-`initialize-instance`, which is called by `make-instance`, just after
-it has created a new instance but didn't initialize it yet with the
+To further control the creation of object instances, we can specialize the method
+`initialize-instance`. It is called by `make-instance`, just after
+a new instance was created but wasn't initialized yet with the
 default initargs and initforms.
 
 It is recommended (Keene) to create an after method, since creating a
 primary method would prevent slots' initialization.
 
 ~~~lisp
-(defmethod initialize-instance :after ((obj person) &key)
+(defmethod initialize-instance :after ((obj person) &key) ;; note &key
   (do something with obj))
 ~~~
 
-Another rational. The CLOS implementation of
+A typical example would be to validate the initial values. Here we'll
+check that the person's name is longer than 3 characters:
+
+~~~lisp
+(defmethod initialize-instance :after ((obj person) &key)
+  (with-slots (name) obj
+    (assert (>= (length name) 3))))
+~~~
+
+So this call doesn't work anymore:
+
+~~~lisp
+(make-instance 'person :name "me" )
+;; The assertion (>= #1=(LENGTH NAME) 3) failed with #1# = 2.
+;;   [Condition of type SIMPLE-ERROR]
+~~~
+
+We are prompted into the interactive debugger and we are given a
+choice of restarts (continue, retry, abort).
+
+So while we're at it, here's an assertion that uses the debugger
+features to offer to change "name":
+
+~~~lisp
+(defmethod INITIALIZE-INSTANCE :after ((obj person) &key)
+  (with-slots (name) obj
+    (assert (>= (length name) 3)
+            (name)  ;; creates a restart that offers to change "name"
```

Comment:
Huh, didn't know that was how you did that.

Reply:
yeah, thanks to this video: * [Video tutorial: introduction on conditions and restarts](http://nklein.com/2011/03/tutorial-introduction-to-conditions-and-restarts/), by Patrick Stein.